lib/connectionProvider.js

```diff
@@ -0,0 +1,25 @@
+const { isFunction, cloneDeep } = require('lodash');
+
+class ConnectionConfigProvider {
+  constructor(client) {
+    this.client = client;
+    this.resolvedConnectionSettings = null;
```

Comment:
Why not resolve settings at this step already? At the very least we can definitely do that if it's a static config.